modules/core/test/test_intrin_utils.hpp

```diff
@@ -2029,29 +2030,40 @@ void test_hal_intrin_float16()
 {
     DUMP_ENTRY(v_float16);
 #if CV_FP16
-    TheTest<v_float32>()
-        .test_loadstore_fp16_f32()
+    TheTest<v_float32>().test_loadstore_fp16_f32();
 #if CV_SIMD_FP16
+    TheTest<v_float16>()
         .test_loadstore_fp16()
         .test_float_cvt_fp16()
+        .test_interleave()
+        .test_addsub()
+        .test_mul()
+        // .test_div() // wrong results
```

Comment:
I found this error of wrong results is due to `EXPECT_EQ(..., resC[i])` gets 0 from `resC[i]` but actually it has non-zero value. Maybe `EXPECT_EQ` has problems with fp16? Do you have any insight on this problem? @vpisarev @opencv-alalek @asmorkalov 

Helpful printings from `test_div()`:

```
i = 0, a = 1.000000, b = 8.000000, c = 0.125000
i = 1, a = 2.000000, b = 7.000000, c = 0.285645
/workspace/fp16_bf16_arithm/opencv/modules/core/test/test_intrin_utils.hpp:552: Failure
Expected equality of these values:
  dataA[i] / dataB[i]
    Which is: 0.285714
  resC[i]
    Which is: 0
Google Test trace:
/workspace/fp16_bf16_arithm/opencv/modules/core/test/test_intrin_utils.hpp:550: i=1
i = 2, a = 3.000000, b = 6.000000, c = 0.500000
i = 3, a = 4.000000, b = 5.000000, c = 0.799805
/workspace/fp16_bf16_arithm/opencv/modules/core/test/test_intrin_utils.hpp:552: Failure
Expected equality of these values:
  dataA[i] / dataB[i]
    Which is: 0.8
  resC[i]
    Which is: 0
Google Test trace:
/workspace/fp16_bf16_arithm/opencv/modules/core/test/test_intrin_utils.hpp:550: i=3
i = 4, a = 5.000000, b = 4.000000, c = 1.250000
i = 5, a = 6.000000, b = 3.000000, c = 2.000000
i = 6, a = 7.000000, b = 2.000000, c = 3.500000
i = 7, a = 8.000000, b = 1.000000, c = 8.000000
```

Get printings from `test_div()`:

```.diff
diff --git a/modules/core/test/test_intrin_utils.hpp b/modules/core/test/test_intrin_utils.hpp
index b7f919f6af..b8b37fee53 100644
--- a/modules/core/test/test_intrin_utils.hpp
+++ b/modules/core/test/test_intrin_utils.hpp
@@ -548,6 +548,7 @@ template<typename R> struct TheTest
         for (int i = 0; i < VTraits<R>::vlanes(); ++i)
         {
             SCOPED_TRACE(cv::format("i=%d", i));
+            printf("i = %d, a = %f, b = %f, c = %f\n", i, dataA[i], dataB[i], resC[i]);
             EXPECT_EQ(dataA[i] / dataB[i], resC[i]);
         }
 
@@ -2038,7 +2039,7 @@ void test_hal_intrin_float16()
         .test_interleave()
         .test_addsub()
         .test_mul()
-        // .test_div() // wrong results
+        .test_div() // wrong results
         // .test_abs() // compile error
         .test_cmp()
         .test_sqrt_abs()
```

Reply:
Looks like `EXPECT_EQ` has problem with fp16 value since it gives non-zero value after I make a type conversion with `float()`. But now it comes to accuracy problem that maybe we need to take care of specifically for fp16.

Helpful printings:

```
i = 0, a = 1.000000, b = 8.000000, a / b = 0.125000, c = 0.125000
i = 1, a = 2.000000, b = 7.000000, a / b = 0.285714, c = 0.285645
/Users/fengyuantao/Workspace/fytao/workspace/fp16_bf16_arithm/opencv/modules/core/test/test_intrin_utils.hpp:552: Failure
Expected equality of these values:
  float(dataA[i] / dataB[i])
    Which is: 0.285714
  float(resC[i])
    Which is: 0.285645
Google Test trace:
/Users/fengyuantao/Workspace/fytao/workspace/fp16_bf16_arithm/opencv/modules/core/test/test_intrin_utils.hpp:550: i=1
i = 2, a = 3.000000, b = 6.000000, a / b = 0.500000, c = 0.500000
i = 3, a = 4.000000, b = 5.000000, a / b = 0.800000, c = 0.799805
/Users/fengyuantao/Workspace/fytao/workspace/fp16_bf16_arithm/opencv/modules/core/test/test_intrin_utils.hpp:552: Failure
Expected equality of these values:
  float(dataA[i] / dataB[i])
    Which is: 0.8
  float(resC[i])
    Which is: 0.799805
Google Test trace:
/Users/fengyuantao/Workspace/fytao/workspace/fp16_bf16_arithm/opencv/modules/core/test/test_intrin_utils.hpp:550: i=3
i = 4, a = 5.000000, b = 4.000000, a / b = 1.250000, c = 1.250000
i = 5, a = 6.000000, b = 3.000000, a / b = 2.000000, c = 2.000000
i = 6, a = 7.000000, b = 2.000000, a / b = 3.500000, c = 3.500000
i = 7, a = 8.000000, b = 1.000000, a / b = 8.000000, c = 8.000000
```

Get printings:

```.diff
diff --git a/modules/core/test/test_intrin_utils.hpp b/modules/core/test/test_intrin_utils.hpp
index b7f919f6af..b3dd482c0a 100644
--- a/modules/core/test/test_intrin_utils.hpp
+++ b/modules/core/test/test_intrin_utils.hpp
@@ -548,7 +548,8 @@ template<typename R> struct TheTest
         for (int i = 0; i < VTraits<R>::vlanes(); ++i)
         {
             SCOPED_TRACE(cv::format("i=%d", i));
-            EXPECT_EQ(dataA[i] / dataB[i], resC[i]);
+            printf("i = %d, a = %f, b = %f, a / b = %f, c = %f\n", i, dataA[i], dataB[i], dataA[i] / dataB[i], resC[i]);
+            EXPECT_EQ(float(dataA[i] / dataB[i]), float(resC[i]));
         }
 
         return *this;
@@ -2038,7 +2039,7 @@ void test_hal_intrin_float16()
         .test_interleave()
         .test_addsub()
         .test_mul()
-        // .test_div() // wrong results
+        .test_div() // wrong results
```